### PR TITLE
Add custom metric support

### LINKED ISSUE #13: Target tracking alarms use ProvisionedConcurrencyUtilization with wrong statistic
Hi,

I found that the created target tracking alarms use the ProvisionedConcurrencyUtilization metric with an "average" statistic. In the [Lambda documentation](https://docs.aws.amazon.com/lambda/latest/dg/monitoring-metrics.html) it is clearly stated, that the concurrency metrics have to be viewed with the "maximum" statistic to evaluate if one is hitting concurrency limits.

I understand that this metric is automatically created by using the predefined LambdaProvisionedConcurrencyUtilization metric provided by AWS application auto scaling.

With the statistic set to "average" it is not possible for auto scaling to work correctly resulting in under provisioning  which leads to constant cold starts of lambda functions. Observable with the provisioned concurrency spillover metric.

I cannot understand how with the "average" statistic the auto scaling ever worked and this feels really odd to me. Am I missing something?

### PR DESCRIPTION
**Pull request by fxgtt**
Fixes #13 by adding support for a custom metric.

### CODE REVIEW

**Comment by fxgtt** [comment]
I had an issue with the current types. I had to satisfy the type checker because in the current types `customMetric.dimensions` might be `undefined` even though this cannot be the case at this point:

    (customMetric.dimensions || []).map(d => ({ Name: d.name, Value: d.value }))

I propose to have different types for the input of the config and the created `data` with filled in defaults. The type for the input would be the same as now with most of the values being optional. But the members of `data` should mostly not be optional anymore.

This is a bigger change though and should probably be solved in its own PR.

File: src/plugin.ts [hunk 1]
```diff
@@ -51,6 +55,28 @@ export default class Plugin {
...
+        name: "Resource",
+        value: `${functionName}:provisioned`,
+      }
+    ]
+
+    return {
+      dimensions: customMetric.dimensions || defaultDimensions,
+      metricName: customMetric.metricName || 'ProvisionedConcurrencyUtilization',
+      namespace: customMetric.namespace || 'AWS/Lambda',
+      statistic: customMetric.statistic || 'Average',
```
Comment (claydanford) [inline src/plugin.ts:78]: @fxgtt, thoughts on making this 'Maximum' by default?

File: src/aws/policy.ts [hunk 2]
```diff
@@ -15,10 +15,14 @@ export default class Policy {
     this.name = new Name(options)
   }
 
-  toJSON(): Record<string, unknown> {
+  toJSON(): Record<string, any> {
```
Comment (claydanford) [inline src/aws/policy.ts:18]: Can you disable eslint no-explicit-any on this next line?

`// eslint-disable-next-line @typescript-eslint/no-explicit-any`

File: README.md [hunk 3]
```diff
@@ -25,7 +25,9 @@ plugins:
 
 ## Configuration
```
Comment (claydanford) [inline README.md:26]: Can you extend the configuration example with:
* Minimum predefinedMetric
* Full with predefinedMetric
* Minimum with customMetric
* Full with customMetric

**Review by claydanford — CHANGES_REQUESTED** [review]

**Comment by claydanford** [comment]
@fxgtt FYI this is really great!

**Comment by claydanford** [comment]
@fxgtt, After further consideration I think we can probably remove most of this configuration. I know in #13 we discussed giving full options, and I apologize for any extra work you did. I think to keep it simple, and within the scope of the plugin, we should limit it to just the statistic as you said. You can leave in all the other code for defaults. There really isn't a use case for modifying the other configuration within the `provisionedConcurrency` scope.

What got me onto this, is when I tried to deploy with the changes and could not match the function name correctly under dimensions. Having those options available created problems that can be solved by simply removing the option all together.

What do you think?

File: src/plugin.ts (hunk 1 above)
Comment (fxgtt) [reply src/plugin.ts:78]: Yes, that's probably better. Makes the reason for using customMetric more obvious.

**Comment by fxgtt** [comment]
I was thinking the same about only exposing statistic. There might be use cases if you have your own metric on which you want to scale differently. For example to react earlier than you could with ProvisionedConcurrencyUtilization. But keeping it simple and reducing accidental misconfiguration should be the right way to go for now.

If I understood you correctly, I keep the code as it is and just remove the possibility of configuration?

**Comment by claydanford** [comment]
@fxgtt You got it! We're almost there. There is only two things left that would really button this up.

1. The statistic should be case insensitive and could use the `ucFirst()` utility function to verify it is the correct case. This one isn't required but would be nice.

2. The dimensions being used in your customMetric changes are just the function name. SLS dynamically creates the function name with `{service}-{stage}-{functionName}`. The current implementation would actually not scale at all since the dimensions do not exist to scale against. (let me know if this one needs clarification)

**Comment by fxgtt** [comment]
@claydanford I cannot confirm that the wrong dimension are being created. The created dimensions have `{service}-{stage}-{functionName}` format. If that is what you meant?

**Comment by claydanford** [comment]
> @claydanford I cannot confirm that the wrong dimension are being created. The created dimensions have `{service}-{stage}-{functionName}` format. If that is what you meant?

@fxgtt, the created dimensions, with customMetric, use just the function name as you have written, but sls actually creates the function with the service, stage, and function name in the name of the function. The dimensions of the `FunctionName` and `Resource` are incorrect since it just takes one of those parameters(functionName).

**Comment by fxgtt** [comment]
@claydanford I understood that. But for me it takes service, stage and function name for the created dimension. Not only functionName.

I use `config.name` for that which holds the name from `Serverless.FunctionDefinition`. If I use `config.function` I only get `functionName`.

For me it looks fine.

**Comment by claydanford** [comment]
@fxgtt, Disregard number 2. It was my mistake. This is solid.

**Review by claydanford — APPROVED** [review]

**Comment by claydanford** [comment]
@hemanthj Approved

**Comment by fxgtt** [comment]
Thank you for the quick merge!

**Pull request merged**